Данный ноутбук о нормализации строк, для сравнения аугментированной строки и оригинальной (не актуально)

In [8]:
from langchain_google_vertexai import ChatVertexAI
import os
import json
import re

from typing import Set


In [2]:
PATH = "..\\src\\data\\komus\\dataset.json"

In [3]:
with open(PATH, "r", encoding="UTF-8") as file:
    data = json.load(file)

In [4]:
print(len(data))

83755


In [5]:
for i, product in enumerate(data):
    title = product["title"]
    if "." in title:
        print(i, title)

24 Карандаш механический желтый Attache Economy 0.7 мм
27 Карандаш механический Attache 0.5 мм
30 Карандаш механический серый Erich Krause Megapolis Concept 0.5 мм
33 Карандаш механический черный Pilot H-185 0.5 мм
37 Карандаш механический Attache Selection 0.5 мм
40 Карандаш механический Deli 0.5 мм (S003)
43 Карандаш механический оранжевый Attache Graphix 0.7 мм
47 Карандаш механический M&G 0.5 мм (AMP101781105AWC)
50 Карандаш механический черный Pilot H-187 0.7 мм
53 Карандаш механический синий Pilot H-185 0.5 мм
57 Карандаш механический синий Pilot H-187 0.7 мм
59 Стержень микрографический HB 0.5 мм Pilot PPL-5 (12 грифелей)
61 Карандаш механический Attache Selection Murano 0.7 мм
67 Карандаш механический черный Pilot H-165 0.5 мм
70 Карандаш механический Deli EU998 Arris 0.7 мм
73 Карандаш механический зеленый металлик Faber-Castell Grip1345 0.5 мм
76 Стержень микрографический HB 0.7 мм Pilot PPL-7 (12 грифелей)
78 Карандаш механический синий Pilot 0.5 мм
81 Карандаш механический 

In [14]:
data[:10]

[{'link': 'https://www.komus.ru/katalog/ruchki-karandashi-markery/-karandashi/karandashi-chernografitnye/karandash-chernografitnyj-plastikovyj-nv-s-lastikom-vybor-est-zatochennyj-shestigrannyj-chernyj-korpus-12-shtuk-v-nabore-/p/2113673/',
  'title': 'Карандаш чернографитный пластиковый НВ с ластиком Выбор есть заточенный шестигранный черный корпус (12 штук в наборе)',
  'attributes': {'Торговая марка': 'Выбор есть',
   'Твердость грифеля': 'HB',
   'Наличие ластика': 'Да',
   'Заточенный': 'Да',
   'Профиль карандаша': 'шестигранный',
   'Материал корпуса': 'пластик',
   'Цвет корпуса': 'черный',
   'Длина корпуса карандаша': '185 мм',
   'Единица продажи': 'упаковка',
   'Кол-во штук в единице продажи': '12',
   'Кол-во единиц продаж в промежуточной упаковке (кратность)': '6',
   'Количество единиц продаж в транспортной упаковке': '90',
   'Страна происхождения': 'Россия'},
  'category': ['Карандаши']},
 {'link': 'https://www.komus.ru/katalog/ruchki-karandashi-markery/-karandashi/kar

In [6]:
product = data[77324]
product

{'link': 'https://www.komus.ru/katalog/posuda-i-tekstil/odnorazovaya-posuda-i-upakovka-/pishhevaya-upakovka/upakovka-dlya-fast-fuda/upakovka-bum-osq-burger-l-123-132-kh123-142-kh70mm-kraft-240sht-up-/p/1876041/',
 'title': 'Упаковка бум. OSQ BURGER L 123(132)х123(142)х70мм,крафт,(240шт/уп)',
 'attributes': {'Торговая марка': 'OSQgroup',
  'Тип': 'коробка',
  'Назначение': 'для бургеров',
  'Материал': 'бумага',
  'Разогрев в микроволновке': 'допускается',
  'Цвет материала': 'коричневый',
  'Количество штук в упаковке': '240',
  'Количество единиц продаж в транспортной упаковке': '1',
  'Страна происхождения': 'Россия',
  'Длина': '123/132 мм',
  'Ширина': '123/142 мм',
  'Высота': '70 мм'},
 'category': ['Упаковка для фаст-фуда']}

In [7]:
PROJECT_ID = 'axial-chemist-425510-p2'
LOCATION = "europe-west4"

# Укажите путь к вашему JSON-файлу с кредами
SERVICE_ACCOUNT_FILE = "../secrets/axial-chemist-425510-p2-b4eb1d622fbe.json"
os.environ['GOOGLE_APPLICATION_CREDENTIALS'] = SERVICE_ACCOUNT_FILE


llm_params = {
    'model_name': 'gemini-2.0-flash-001',
    'project': PROJECT_ID,
    'location': LOCATION
}

In [9]:
PROMPT = f"""
Нормализуй данную строку {product["title"]}. 
1) Расскрой все сокращения. 
2) Переведи все прилагательные в существительное. 
3) Сделай отступы между смысловыми частями к примеру между числами в размере.
4) Все приведи в слова во множественном числе в единственное.
Выдай результат в виде одного ответа.
"""

In [12]:
llm = ChatVertexAI(**llm_params)

res = llm.invoke(PROMPT)

normalized_title = res.content

In [13]:
normalized_title

'Упаковка бумага. OSQ бургер L 123 (132) x 123 (142) x 70 миллиметр, крафт, (240 штука / упаковка).\n'

## Примеры разложения:

Упаковка бумага, объект OSQ бургер, размер L 123 (132) x 123 (142) x 70 миллиметр, материал крафт, (240 штука/упаковка).  
Упаковка бумага, объединенная система качества бургер, большой, 123 (132) x 123 (142) x 70 миллиметр, крафт, 240 штука / упаковка.  
Упаковка бумага. OSQ бургер L 123 (132) x 123 (142) x 70 миллиметр, крафт, (240 штука / упаковка).  

## Функции для получения метрик схожести.

### Перекрытие терминов.

In [ ]:
def preprocess_text(text: str) -> Set[str]:
    """
    Предобработка текста: удаление знаков препинания, приведение к нижнему регистру,
    разбиение на слова и возврат множества уникальных терминов.
    """
    # Удаляем знаки препинания и приводим текст к нижнему регистру
    text = re.sub(r'[^\w\s]', '', text.lower())
    # Разбиваем текст на слова
    words = text.split()
    # Возвращаем множество уникальных слов
    return set(words)

def term_overlap_metric(text1: str, text2: str) -> float:
    """
    Вычисление метрики перекрытия терминов между двумя текстами.
    
    :param text1: Первый текст
    :param text2: Второй текст
    :return: Значение метрики перекрытия терминов (от 0 до 1)
    """
    # Получаем множества уникальных терминов для обоих текстов
    terms1 = preprocess_text(text1)
    terms2 = preprocess_text(text2)
    
    # Находим пересечение множеств (общие термины)
    common_terms = terms1.intersection(terms2)
    
    # Находим объединение множеств (все уникальные термины)
    all_terms = terms1.union(terms2)
    
    # Вычисляем метрику как отношение размера пересечения к размеру объединения
    if len(all_terms) == 0:
        # Если оба текста пустые, возвращаем 0
        return 0.0
    return len(common_terms) / len(all_terms)